### ISSUE TITLE
[4.0.5][Feature request] Ability to disable/change the default orderBy primary key in ListOperation

### ISSUE THREAD

**Issue by jsvini** [Laravel-Backpack/CRUD#2112]
# Bug report

### What I did

The ListOperation always add an order by primary key to the query:

https://github.com/Laravel-Backpack/CRUD/blob/4.0.5/src/app/Http/Controllers/Operations/ListOperation.php#L114

I think that should be an option to disable this or at least to manual set the column name, because this makes impossible to do a join to another table that have the same column name, you get this error in MySQL:

`Integrity constraint violation: 1052 Column 'id' in order clause is ambiguous`

Because both tables in the join have the same `'id'` column.

If we add the table name in query instead of just the column name this issue could be solved.

Instead of:
... order by `id` desc limit 10;

it could be:
... order by `table_name`.`id` desc limit 10;

---------

This piece below in `ListOperation.php` ignores the primary column orderBy if we already add a custom order to it, but it doesn't catch if we add the column name like this:

$this->crud->orderBy( 'table_name.id');

only if we write like this:

$this->crud->orderBy( 'id');

https://github.com/Laravel-Backpack/CRUD/blob/4.0.5/src/app/Http/Controllers/Operations/ListOperation.php#L116

**Comment by ckdot** [comment]
+1

**Comment by pxpm** [comment]
Hello @jsvini and @ckdot 

I think i have the solution (at least for one of the problems), but i would like you to test it before submiting a PR or show me your code so i could test it. 

`In Traits\Query.php function orderBy`

replace: 

``` return $this->query->orderBy($field, $order); ```

with:

``` return $this->query->orderBy($this->model->getTable().'.'.$field, $order); ```

This will force crud to append the table name to all orderBy. But i think that's not the complete solution, can you guys provide me some sample code so i could test it out ?

Best,
Pedro

**Comment by tabacitu** [comment]
Thank you for reporting this @jsvini . Does sound like a bug to me. My bad, of course!

@pxpm I think your solution should fix the error too. But I haven't tested it. 

To further prevent frustration on this topic, I think we could  add a new setting like ```list.orderByPrimaryKey```with the options ```'ASC', 'DESC', false``` (default ```'DESC'```). So that people could, if they wanted, change the behaviour, or disable it entirely. That sounds like a good long-term solution that would allow the maximum flexibility right @jsvini , @ckdot , @pxpm ?

Cheers!

**Comment by pxpm** [comment]
Hey @tabacitu 

Not sure the real problem is with desc/asc. Disabling it would ofcourse work. 

But developers might not want to disable the order feature, I think the problem is that crud only checks if the passed column is model key:

From ListOperation:
```
...
if ($item['column'] == $this->crud->model->getKeyName()) {
                $hasOrderByPrimaryKey = true;

                return false;
            }
...
``` 

Maybe we can write this like:

```
if ($item['column'] == $this->crud->model->getKeyName() || $item['column'] == $this->crud->model->getTable().'.'.$this->crud->model->getKeyName()) {
                $hasOrderByPrimaryKey = true;

                return false;
            }
```
But if developer orderBy('some_other_table.field') this check will fail. 

So my question is, it's mandatory that the model is always ordered by primary key? or this code is misleading and that variable should read `$hasDeveloperOrder` ?

In that case we just need to check if `isset($field['column'])` if it's not set we apply default order by primary key.

I think the solution lies around this lines. Let me know your thoughts.

Best,
Pedro

**Comment by tabacitu** [comment]
I agree, the disable/asc/desc setting would NOT be for solving this issue, but for added flexibility. And providing a workaround if/when developers encounter other issues with our default orderBy logic.

So the possible changes for this area of the code would be:
(1) Fix this bug. Make it so the ID columns don't conflict when adding another ```orderBy('category.id')``` clause. I don't think ```if ($item['column'] == $this->crud->model->getKeyName() || $item['column'] == $this->crud->model->getTable().'.'.$this->crud->model->getKeyName()) {``` could be a fix as it could still be true when someone adds an additional order, for some other table. Maybe a more strict ```if ($item['column'] == $this->crud->model->getTable().'.'.$this->crud->model->getKeyName()) {``` will work better, in addition to the default orderBy using the explicit table name too, either by using ```orderByRaw()``` or ```orderBy``` classic (not sure you _can_ specify the table name with the classic orderBy).

(2) Make this area more flexible. Add a setting to disable (or change) the default order. I agree with you, I'm not sure _anybody_ would _ever_ change the order from DESC to ASC. Why would they? It's more intuitive to have last items first in the table. But I thought that disabling the feature is a must - so we'd add a setting for that. And then, if we already have a setting for it, why not make the setting more general, and allow developers to change the order to DESC to? 😄 I'm not adamant about this, if you think true/false is better go with that. 

Cheers!

**Comment by pxpm** [comment]
I will give it a deeper look ;)

Thanks for feedback @tabacitu .

Best,
Pedro

**Comment by pxpm** [comment]
So back on this... here are my findings:

@tabacitu was right, 'standard' `orderBy` does not allow the `table.field` notation. 

So without any changes, the current query would produce an ambiguous column in sql:

```
public function setupListOperation()
    {

        $this->crud->query->join('articles','monsters.select2_from_ajax','=','articles.id')->orderByRaw('articles.title ASC');
}
```
`"SQLSTATE[23000]: Integrity constraint violation: 1052 Column 'id' in order clause is ambiguous (SQL: select * from `monsters` inner join `articles` on `monsters`.`select2_from_ajax` = `articles`.`id` order by articles.title ASC, `id` desc limit 10)"`

However changing the `orderByDesc` in [ListOperation.php](https://github.com/Laravel-Backpack/CRUD/blob/9b611865c41143f8ebf2ab39be7af3b564bb0f8d/src/app/Http/Controllers/Operations/ListOperation.php#L127) to:

```
$crudOrderDirection = $this->crud->getOperationSetting('orderingDirection');

...

if (! $hasOrderByPrimaryKey && $crudOrderDirection != false) {
            $this->crud->query->orderByRaw($this->crud->model->getTable() . '.' . $this->crud->model->getKeyName(), $crudOrderDirection);
        }
```
Will run the expected query:

` select * from `monsters` inner join `articles` on `monsters`.`select2_from_ajax` = `articles`.`id` order by articles.title ASC, monsters.id limit 10 `

We also need to take into account the [datatable column ordering](https://github.com/Laravel-Backpack/CRUD/blob/9b611865c41143f8ebf2ab39be7af3b564bb0f8d/src/app/Http/Controllers/Operations/ListOperation.php#L101)

Needs to change to: 

` $this->crud->query->orderByRaw($this->crud->model->getTable() . '.' . $column['name'] . ' ' . $column_direction);
`

I don't forecast any breaking apps, we are just appending the table name to the query, but well I could be wrong. 

It would help alot if @jsvini and @ckdot could test this on the projects where they are having issues. 

Just change line 101 and 127 of `src/app/Http/Controllers/Operations/ListOperation.php` with:

101 - `$this->crud->query->orderByRaw($this->crud->model->getTable() . '.' . $column['name'] . ' ' . $column_direction);`

127 - ` $this->crud->query->orderByRaw($this->crud->model->getTable() . '.' . $this->crud->model->getKeyName() . ' DESC');`

Let me know how it goes in your scenarios.

Best,
Pedro

**Comment by acewebsa** [comment]
I was having this issue and can confirm that pxpm's solution with changing lines 101 and 127 in `src/app/Http/Controllers/Operations/ListOperation.php` fixed the issue.

**Comment by tabacitu** [comment]
@pxpm I believe PR #3015 fixes this too, right? If so let's close this issue and move the conversation there, please. 
If anybody else can confirm #3015 fixes the problem please let us know - it's make for a faster merge.

Cheers!
